### PR TITLE
Configure Item submission process at community level

### PR DESCRIPTION
**Pull request by toniprieto**
## Description
This PR allows to use a community handle to map the descendant collections to an item submission process.  This can be useful for repositories where top-level communities represent different document types or scopes and descendant collections should share the same submission process.

## Instructions for Reviewers

To test it, create two top-level communities with collections without a entity-type assigned and map the first community handle to a new item submission process in `item-submission.xml` (for example, one with the cclicense). 

```
  <name-map community-handle="<community-handle>" submission-name="test"/>

    <submission-process name="test">
        <step id="collection"/>
        <step id="traditionalpageone"/>
        <step id="upload"/>
        <step id="cclicense"/>
        <step id="license"/>
    </submission-process>
```
This new item submission process should be used for collections descendant of the community configured.

Additional info:
- It should also work with sub-communities handles.
- Mappings assigned by collection-handle or entity-type have a higher priority than this new mapping by community.
- A new test has been added to check this new configuration

## Checklist

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [ ] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

### CODE REVIEW

File: dspace-api/src/main/java/org/dspace/app/util/SubmissionConfigReader.java [hunk 1]
```diff
@@ -210,18 +221,39 @@ public int countSubmissionConfigs() {
...
+            // get the name of the submission process config for this collection
             submitName = collectionToSubmissionConfig
-                .get(DEFAULT_COLLECTION);
+                .get(col.getHandle());
+            if (submitName != null) {
+                return getSubmissionConfigByName(submitName);
+            }
+
+            try {
+                List<Community> communities = col.getCommunities();
```
Comment (abollini) [inline dspace-api/src/main/java/org/dspace/app/util/SubmissionConfigReader.java:242]: can we optimize the performance here for institution that would not use the new feature?
we should check if the `communityToSubmissionConfig` map is not empty otherwise we can just skip traversing the comm/coll hierarchy

File: dspace-api/src/test/java/org/dspace/app/util/SubmissionConfigTest.java [hunk 2]
```diff
@@ -87,4 +109,35 @@ public void testReadAndProcessTypeBindSubmissionConfig()
...
+
+        // Sep up a structure with one top community and two subcommunities
+        // with one collection
+        when(col1.getHandle()).thenReturn("123456789/not-mapped1");
+        when(col1.getCommunities()).thenReturn(List.of(subcom1));
+
+        when(col2.getHandle()).thenReturn("123456789/not-mapped2");
+        when(col2.getCommunities()).thenReturn(List.of(subcom2));
+
+        when(subcom1.getHandle()).thenReturn("123456789/subcommunity-test");
```
Comment (abollini) [inline dspace-api/src/test/java/org/dspace/app/util/SubmissionConfigTest.java:125]: subcom1 should be a child of topcom to verify that the configuration of the "close" community is used

File: dspace-api/src/test/java/org/dspace/app/util/SubmissionConfigTest.java [hunk 3]
```diff
@@ -87,4 +109,35 @@ public void testReadAndProcessTypeBindSubmissionConfig()
...
+        // for col1, it should return the item submission form defined for their parent subcom1
+        SubmissionConfig submissionConfig1 =
+            new SubmissionConfigReader().getSubmissionConfigByCollection(col1);
+        assertEquals("subcommunitytest", submissionConfig1.getSubmissionName());
+
+        // for col2, it should return the item submission form defined for topcom
+        SubmissionConfig submissionConfig2 =
+            new SubmissionConfigReader().getSubmissionConfigByCollection(col2);
+        assertEquals("topcommunitytest", submissionConfig2.getSubmissionName());
+
```
Comment (abollini) [inline dspace-api/src/test/java/org/dspace/app/util/SubmissionConfigTest.java:141]: should we check that a collection with a dedicated submission form within the topcomm get the collection specific form instead than the topcomm one?

Comment (abollini) [reply dspace-api/src/test/java/org/dspace/app/util/SubmissionConfigTest.java:141]: we also need to test the respect of the precedence of the entity mapping over the community mapping

**Review by abollini — DISMISSED** [review]
hi @toniprieto thanks for your contribution, changes look good to me. We just need some extra test to guarantee us that the implementation is robust against future refactoring. I have also suggested a small performance improvement see comments inline

**Comment by toniprieto** [comment]
@abollini Thanks for the review!

I've addressed part of the feedback - the performance improvement (thanks, good idea!) and testing that a directly mapped collection works correctly - but I'm having issues testing the precedence of entity mapping over community mapping. I have also converted the test into an integration test, I think it is clearer now.

When I add the configuration to add a per-entityType mapping in the item-submssion.xml used in the tests...

```
<name-map collection-entity-type="Publication" submission-name="publicationtest"/>
```

...the service that loads this configuration (`SubmissionConfigService`) is not always initialized because it needs to do a request to the solr search core to get the collections that have assigned this entity type and it seems that its solr core search is not always available during this initialization. I've seen this stack trace debugging this issue:

```
org.dspace.app.util.SubmissionConfigReaderException: Cannot perform a discovery search for Item Submission Configuration
	at org.dspace.app.util.SubmissionConfigReader.buildInputs(SubmissionConfigReader.java:177)
	at org.dspace.app.util.SubmissionConfigReader.<init>(SubmissionConfigReader.java:134)
	at org.dspace.app.util.SubmissionConfigIT.testSubmissionMapByCommunityHandleSubmissionConfig(SubmissionConfigIT.java:62)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:290)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
Caused by: org.dspace.discovery.SearchServiceException: No such core: search
	at org.dspace.discovery.SolrServiceImpl.search(SolrServiceImpl.java:774)
	at org.dspace.content.CollectionServiceImpl.findAllCollectionsByEntityType(CollectionServiceImpl.java:1122)
	at org.dspace.app.util.SubmissionConfigReader.processMap(SubmissionConfigReader.java:441)
	at org.dspace.app.util.SubmissionConfigReader.doNodes(SubmissionConfigReader.java:378)
	at org.dspace.app.util.SubmissionConfigReader.buildInputs(SubmissionConfigReader.java:172)
	... 34 more
Caused by: org.apache.solr.common.SolrException: No such core: search
	at org.apache.solr.client.solrj.embedded.EmbeddedSolrServer.request(EmbeddedSolrServer.java:197)
	at org.apache.solr.client.solrj.SolrRequest.process(SolrRequest.java:214)
	at org.apache.solr.client.solrj.SolrClient.query(SolrClient.java:1035)
	at org.apache.solr.client.solrj.SolrClient.query(SolrClient.java:1051)
	at org.dspace.discovery.SolrServiceImpl.retrieveResult(SolrServiceImpl.java:995)
	at org.dspace.discovery.SolrServiceImpl.search(SolrServiceImpl.java:771)
	... 38 more
```
The new test works correctly if it is run alone but there are errors if all of them are run. I'm not sure how to solve this issue and it doesn't seem related to the change made in this PR.

File: dspace-api/src/test/java/org/dspace/app/util/SubmissionConfigIT.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+ */
+public class SubmissionConfigIT extends AbstractIntegrationTestWithDatabase {
+
+    SubmissionConfigService submissionConfigService;
+
+    @Before
+    @Override
+    public void setUp() throws Exception {
+        super.setUp();
+        submissionConfigService = SubmissionServiceFactory.getInstance().getSubmissionConfigService();
```
Comment (tdonohue) [inline dspace-api/src/test/java/org/dspace/app/util/SubmissionConfigIT.java:35]: I'd recommend deleting this entire `setUp()` method and simply calling  `SubmissionServiceFactory.getInstance().getSubmissionConfigService();` down in the `@Test` method below _when you need it._

I'm worried that calling this in the setUp may result in this getting called **before** the rest of the test database & other services initialize.

File: dspace-api/src/test/java/org/dspace/builder/AbstractBuilder.java [hunk 5]
```diff
@@ -178,7 +178,12 @@ public static void init() {
         systemWideAlertService = DSpaceServicesFactory.getInstance().getServiceManager()
                                                       .getServicesByType(SystemWideAlertService.class).get(0);
         try {
-            submissionConfigService = SubmissionServiceFactory.getInstance().getSubmissionConfigService();
+            SubmissionServiceFactory submissionServiceFactory = SubmissionServiceFactory.getInstance();
+            if (submissionServiceFactory != null) {
+                submissionConfigService = submissionServiceFactory.getSubmissionConfigService();
+            } else {
+                log.error("ERROR: Error initializing submissionServiceFactory");
```
Comment (tdonohue) [inline dspace-api/src/test/java/org/dspace/builder/AbstractBuilder.java:185]: I find the changes to this file to be odd.  I don't understand why this small PR suddenly requires extra error handling here.  That almost implies to me that we are missing something & we have an unintended side effect to how the SubmissionConfigService is loading.

**Review by tdonohue — COMMENTED** [review]
@toniprieto : Gave this a look today and a test.  First, the test shows this is working properly.  But, obviously you seem to have two IT failures in this PR.  It's not completely obvious to me **why** those IT failures are occurring.  It almost seems like the ITs are not initializing properly...and these appear to be the first tests which attempt to use/reload the `SubmissionConfigService`

From the logs of the IT failures...it almost seems to me like that service isn't loading properly, but I cannot figure out why. Unfortunately, because I have a lot of other PR reviews on my plate, I don't have time to dig in further right now. One minor suggestion below to try, but not sure if it will be helpful.

It would really be nice to either find a way to get these tests to succeed **or** find the root cause.  Otherwise, I'm worried that there may be unintended consequences to the changes in this PR.

Basically, I'm +1 this PR _if we can get the tests to succeed._

File: dspace-api/src/test/java/org/dspace/app/util/SubmissionConfigIT.java (hunk 4 above)
Comment (toniprieto) [reply dspace-api/src/test/java/org/dspace/app/util/SubmissionConfigIT.java]: Thanks for the suggestion, I've tried it but it hasn't worked either.

**Review by tdonohue — APPROVED** [review]
👍 Thanks @toniprieto !  I think this is "good enough".  I've manually tested all these scenarios to verify it's working properly:
1. Setting a submission form at a Community inherits to a child collection (provided the child collection has no specified submission form)
2. If the child collection specifies an Entity Type, this overrides any submission form configured for a parent community
3. If a child collection has its own configuration in `item-submission.xml`, this also overrides any submission form configured for a parent community

Code looks good & the tests look good enough for now.

**Comment by toniprieto** [comment]
@tdonohue can you give me permissions to edit this page/wiki? I'll add the documentation in the upcoming weeks.

**Comment by tdonohue** [comment]
@toniprieto : I've just updated your wiki permissions for your `tonijpj` account.  You should now have full edit permissions across the DSpace documentation.  Thanks!

**Comment by toniprieto** [comment]
I forgot to comment here, I added the documentation a few weeks ago: https://wiki.lyrasis.org/display/DSDOC8x/Submission+User+Interface#SubmissionUserInterface-AssigningacustomSubmissionProcesstoaCollection
